File: kotlin2/BattleSim.kt
// P111: Kotlin v2 - Coroutine-based AI
import kotlinx.coroutines.*

data class Unit(val id: Long, val health: Float, val damage: Float, val x: Float, val y: Float)

class BattleSimulator {
    private val units = mutableListOf<Unit>()
    
    fun addUnit(unit: Unit) = units.add(unit)
    
    fun calculatePower(): Float = units.sumOf { (it.health * it.damage).toDouble() }.toFloat() / 100f
    
    suspend fun analyzeAllUnits(): List<Unit> = withContext(Dispatchers.Default) {
        units.map { it.copy() }
    }
    
    fun findThreats(): Map<Long, Float> {
        val threats = mutableMapOf<Long, Float>()
        for (u in units) {
            val nearby = units.count { distance(u, it) < 50f && it.id != u.id }
            threats[u.id] = nearby * 10f
        }
        return threats
    }
    
    private fun distance(a: Unit, b: Unit): Float {
        val dx = a.x - b.x
        val dy = a.y - b.y
        return kotlin.math.sqrt(dx * dx + dy * dy)
    }
}

fun main() = runBlocking {
    val sim = BattleSimulator()
    sim.addUnit(Unit(1, 40f, 5f, 10f, 10f))
    println("Power: ${sim.calculatePower()}")
}

File: scala2/src/main/scala/BattleSim.scala
// P118: Scala v2 - Akka-based Actor System
import akka.actor._

case class Unit(id: Int, health: Float, damage: Float, x: Float, y: Float)
case class CalculatePower()
case class FindThreats()

class BattleSimulator extends Actor {
    var units: List[Unit] = List()
    
    def receive: Receive = {
        case u: Unit => units = units :+ u
        case CalculatePower() => 
            val power = units.map(u => u.health * u.damage).sum / 100.0f
            sender() ! power
        case FindThreats() =>
            val threats = units.map { u =>
                val nearby = units.count(e => e.id != u.id && distance(u, e) < 50f)
                (u.id, nearby * 10f)
            }.toMap
            sender() ! threats
    }
    
    def distance(a: Unit, b: Unit): Float = {
        val dx = a.x - b.x
        val dy = a.y - b.y
        Math.sqrt(dx*dx + dy*dy).toFloat
    }
}

object Main extends App {
    val system = ActorSystem("SC2Game")
    val sim = system.actorOf(Props[BattleSimulator](), "battleSim")
    sim ! Unit(1, 40, 5, 10, 10)
    sim ! CalculatePower()
}
